### PR TITLE
Speech Command Recognition With Tensorflow.js and React.js

### LINKED ISSUE #1693: Getting started with Speech Command Recognition With Tensorflow.JS and React.JS
### Proposed title of the article
Getting started with Speech Command Recognition With Tensorflow.JS and React.JS 

### Introduction paragraph (2-3 paragraphs):
In this tutorial, we will learn how to perform speech command recognition in real-time and demonstrate the result of predictions. The idea is to create a React app and trigger the web browser system microphone that feeds the audio data. Then, we are going to load the speech-commands model provided by [TensorFlow.org](http://tensorflow.org/). Feeding the audio data to the model neural network, we are going to listen to the audio and make the predictions. Lastly, we are also going to improve the accuracy recognition by applying argmax function to our model.

### Key takeaways: 
- How to access the audio data stream from the browser microphone.
- How to predict the speech commands using a pre-trained speech-commands model from TensorFlow.
- How to improve the overall prediction/recognition probability by using the argmax function.

### PR DESCRIPTION
**Pull request by Worawat101**
### Article checklist
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.

This fixes #1693 

#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Have you placed your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

### CODE REVIEW

File: articles/speech-command-recognition-with-tensorflow-and-react/index.md [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+author: worawat-kaewsanmaung
+date: 2021-03-15T00:00:00-10:00
+topics: []
+excerpt_separator: <!--more-->
+images:
+
+  - url: /engineering-education/speech-command-recognition-with-tensorflow-and-react/hero.png
+    alt: Speech Command Recognition in a React Project with Tensorflow 
+---
+There are many AI-powered computer systems nowadays that can use speech-based communication to communicate with humans. The main logic that is used in these types of systems is the ability to recognize the speech and perform tasks accordingly. In this tutorial, we are going to explore the basics of speech/commands recognition using a speech-commands recognizer model provided by TensorFlow in a React app ecosystem. The tutorial also demonstrates the use of a system microphone for speech audio input and listen to the audio to make probabilistic predictions.
```
Comment (cyberShaw) [inline articles/speech-command-recognition-with-tensorflow-and-react/index.md:17]:
```suggestion
There are many AI-powered computer systems nowadays that can use speech-based communication to communicate with humans. The main logic that is used in these types of systems is the ability to recognize speech and perform tasks accordingly. In this tutorial, we are going to explore the basics of speech/command recognition using a speech-command recognizer model provided by TensorFlow in a React app ecosystem. The tutorial also demonstrates the use of a system microphone for speech audio input and listen to the audio to make probabilistic predictions.
```

File: articles/speech-command-recognition-with-tensorflow-and-react/index.md [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+<!--more-->
+In this tutorial, we will learn how to perform speech command recognition in real-time and demonstrate the result of predictions. Then, we are going to load the speech-commands model provided by TensorFlow.org. Feeding the audio data to the model neural network, we are going to listen to the audio and make the predictions. Lastly, we are also going to improve the accuracy recognition by applying the argmax function to our model.
```
Comment (cyberShaw) [inline articles/speech-command-recognition-with-tensorflow-and-react/index.md:19]:
```suggestion
In this tutorial, we will learn how to perform speech command recognition in real-time and demonstrate the result of predictions. Then, we are going to load the speech-commands model provided by TensorFlow.org. Feeding the audio data to the model neural network, we are going to listen to the audio and make the predictions. Lastly, we are also going to improve the accuracy of recognition by applying the argmax function to our model.
```

File: articles/speech-command-recognition-with-tensorflow-and-react/index.md [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+### What we will cover in this tutorial...
+
+- To access the audio data stream from the browser microphone.
+- To precise, the speech commands using the pre-trained speech-commands model from TensorFlow.
```
Comment (cyberShaw) [inline articles/speech-command-recognition-with-tensorflow-and-react/index.md:24]:
```suggestion
- To be precise, the speech commands using the pre-trained speech-commands model from TensorFlow.
```

File: articles/speech-command-recognition-with-tensorflow-and-react/index.md [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+- To improve the overall prediction/recognition probability by using the argmax function.
+
+*Let's get started!*
+
+### Create React Project
+First, we are going to create a new React App project. For that, we need to run the following command in the required local directory:
+
+```jsx
```
Comment (cyberShaw) [inline articles/speech-command-recognition-with-tensorflow-and-react/index.md:32]:
```suggestion
```bash
```

File: articles/speech-command-recognition-with-tensorflow-and-react/index.md [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+npx create-react-app speech_command
+```
+
+Then, we need to open the project in our code editor IDE (VScode) as shown in the screenshot below:
```
Comment (cyberShaw) [inline articles/speech-command-recognition-with-tensorflow-and-react/index.md:36]:
```suggestion
Then, we need to open the project in our code editor IDE (VSCode) as shown in the screenshot below:
```

File: articles/speech-command-recognition-with-tensorflow-and-react/index.md [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+
+![create new react app](/engineering-education/speech-command-recognition-with-tensorflow-and-react/1-create-new-react-app.png)
+
+Now, we can execute the command  `yarn start` or `npm start` to run the project on the port 3000 of our [localhost](http://localhost):
```
Comment (cyberShaw) [inline articles/speech-command-recognition-with-tensorflow-and-react/index.md:40]:
```suggestion
Now, we can execute the command  `yarn start` or `npm start` to run the project on the port 3000 of our [localhost](http://localhost:3000):
```

File: articles/speech-command-recognition-with-tensorflow-and-react/index.md [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+![preview react app](/engineering-education/speech-command-recognition-with-tensorflow-and-react/2-preview-react-app.png)
+
+### Installing Tensorflow.JS & Speech Command Recognition
+
+Now, we are going to install the main TensorFlow.JS package along with the speech commands recognition package provided by TensorFlow.org. for installation package, we need to execute the following command in our project commmand line or terminal:
```
Comment (cyberShaw) [inline articles/speech-command-recognition-with-tensorflow-and-react/index.md:46]:
```suggestion
Now, we are going to install the main TensorFlow.js package along with the speech commands recognition package provided by TensorFlow.org. For installing the package, we need to execute the following command in our project command line or terminal:
```

File: articles/speech-command-recognition-with-tensorflow-and-react/index.md [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+
+```bash
+yarn add @tensorflow/tfjs @tensorflow-models/speech-commands
+```
+
+- @tensorflow/tfjs: It is the core Tensorflow package based on JavaScript.
+- @tensorflow-models/speech-commands: It is a Tensorflow models that enables us to perform recognition of spoken commands.
```
Comment (cyberShaw) [inline articles/speech-command-recognition-with-tensorflow-and-react/index.md:53]:
```suggestion
- @tensorflow-models/speech-commands: It is a Tensorflow model that enables us to perform recognition of spoken commands.
```

File: articles/speech-command-recognition-with-tensorflow-and-react/index.md [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+After the successful installation of the required packages. we need to import them in the App.js file of our React project as shown in the code snippet below:
+
+```jsx
+import * as tf from "@tensorflow/tfjs"
+import * as speech from "@tensorflow-models/speech-commands"
+```
+
+### Setting Up the Application States
+
+Then, we need to define the state variable for model, action, and labels using the `useState` hook as shown in the code snippet below:
```
Comment (cyberShaw) [inline articles/speech-command-recognition-with-tensorflow-and-react/index.md:66]: Please briefly explain how the useState hook works. This might be useful for a beginner

File: articles/speech-command-recognition-with-tensorflow-and-react/index.md [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+const [labels, setLabels] = useState(null)
+```
+
+They are all initialized to null value at the start.
+
+### Load the Speech Recognizer Model
+
+Here, we are going to create a function called `loadModel` that loads the Speech Recognizer model into our app. First, a `recognizer` is initialized using the `create` method provided by the `speech` module. This loads the speech model in the app. Then, we ensure that the model is loaded using the `ensureModelLoaded` method from the `recognizer`. Then, we set the `recognizer` to `models` state and `wordLabels` in the recognizer to `labels` state. Then, we call the `loadModel` function inside the `useEffect` hook so that it is triggered when the app starts. The implementation is shown in the code snippet below:
+
+```jsx
```
Comment (cyberShaw) [inline articles/speech-command-recognition-with-tensorflow-and-react/index.md:81]: Please add code comments, this would make the code more readable.

File: articles/speech-command-recognition-with-tensorflow-and-react/index.md [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+const [model, setModel] = useState(null)
+const [action, setAction] = useState(null)
+const [labels, setLabels] = useState(null)
+```
+
+They are all initialized to null value at the start.
+
+### Load the Speech Recognizer Model
+
+Here, we are going to create a function called `loadModel` that loads the Speech Recognizer model into our app. First, a `recognizer` is initialized using the `create` method provided by the `speech` module. This loads the speech model in the app. Then, we ensure that the model is loaded using the `ensureModelLoaded` method from the `recognizer`. Then, we set the `recognizer` to `models` state and `wordLabels` in the recognizer to `labels` state. Then, we call the `loadModel` function inside the `useEffect` hook so that it is triggered when the app starts. The implementation is shown in the code snippet below:
```
Comment (cyberShaw) [inline articles/speech-command-recognition-with-tensorflow-and-react/index.md:79]: Please also briefly explain how the useEffect hook works

File: articles/speech-command-recognition-with-tensorflow-and-react/index.md [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+const recognizeCommands = async () => {
+    console.log("Listening for commands");
+    model.listen(
+      (result) => {
+        console.log(result.spectrogram);
+      },
+      { includeSpectrogram: true, probabilityThreshold: 0.9 }
+    );
+    setTimeout(() => model.stopListening(), 10e3);
+  };
```
Comment (cyberShaw) [inline articles/speech-command-recognition-with-tensorflow-and-react/index.md:112]: Please add code comments here too

File: articles/speech-command-recognition-with-tensorflow-and-react/index.md [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+
+Now, we are going to add the `argmax` function to our speech recognizer model to improve its detection to the highest possible result. The actual source code for this function can be found [here](https://gist.github.com/engelen/fbce4476c9e68c52ff7e5c2da5c24a28).
+
+```jsx
+function argMax(arr){
+  return arr.map((x, i) => [x, i]).reduce((r, a) => (a[0] > r[0] ? a : r))[1];
+}
+```
+
+The function uses `map` logic along with `reduce` method to improve the overall probability of detection. Hence, it is a Map-Reduce algorithm.
```
Comment (cyberShaw) [inline articles/speech-command-recognition-with-tensorflow-and-react/index.md:139]: Please explain how `map` and `reduce` works in two lines.

File: articles/speech-command-recognition-with-tensorflow-and-react/index.md [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+
+```jsx
+const recognizeCommands = async () =>{
+  console.log('Listening for commands')
+  model.listen(result=>{
+    setAction(labels[argMax(Object.values(result.scores))])
+  }, {includeSpectrogram:true, probabilityThreshold:0.7})
+  setTimeout(()=>model.stopListening(), 10e3)
+}
+```
```
Comment (cyberShaw) [inline articles/speech-command-recognition-with-tensorflow-and-react/index.md:151]: Please add code comments

**Review by cyberShaw — CHANGES_REQUESTED** [review]
Great article, @Worawat101 !
Really good work. I have some minor suggestions. Please make them so we can move to the final round of review.

**Comment by Worawat101** [comment]
@cyberShaw I already fixed all your comment

**Comment by cyberShaw** [comment]
> @cyberShaw I already fixed all your comment

Great work, I'll do a review.

**Comment by Worawat101** [comment]
@cyberShaw hi I wundering if your help me win this week with complete this task

**Comment by cyberShaw** [comment]
Thank you for making the changes, @Worawat101!
This article is ready for final review.

**Comment by hectorkambow** [comment]
plagiarism check done
